src/ogdf/basic/graph_generators/operations.cpp

```diff
@@ -246,4 +251,83 @@ void rootedProduct(const Graph& G1, const Graph& G2, Graph& product, NodeMap& no
 	});
 }
 
+void complement(Graph& G, bool directional, bool allow_self_loops) {
+	for (node n1 : G.nodes) {
+		for (node n2 : G.nodes) {
+			if (n1->index() > n2->index()) {
+				continue;
+			}
+			if (!allow_self_loops && n1->index() == n2->index()) {
+				continue;
+			}
+
+			edge e = G.searchEdge(n1, n2, true);
+			edge er = G.searchEdge(n2, n1, true);
+
+			if (e) {
+				G.delEdge(e);
+			}
+			if (er) {
+				G.delEdge(er);
+			}
+
+			if (directional) {
+				if (!e) {
+					G.newEdge(n1, n2);
+				}
+				if (!er && !allow_self_loops) {
```

Comment:
why `!allow_self_loops` here?

Reply:
That's a mistake, it should've been `!er && (n1 != n2)` instead. The idea is to prevent a second self loop.